### DISCUSSION TITLE
curl: (22) The requested URL returned error: 400

### DISCUSSION THREAD

**Discussion by akgreat22** [biostars/biostar-handbook-issues#236] [Q&A]
I used the esearch command "esearch -db nucleotide -query PRJNA257197 | efetch -format fasta > genome.fa". below is what I get as one of the error messages; I use Ubuntu on windows 10

-X POST https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi -d db=nucleotide&id=Please%2CUnable%2Cchmod%2Cdwn%2Centrez%2Centrezdirect%2Cexecutable.%2Cexecute%2Cf%2Cfollowing%2Cftp.ncbi.nlm.nih.gov%2Cgunzip%2Clocate%2Cnquire%2Cthe%2Cto%2Cx%2Cxtract%2Cxtract.Linux%2Cxtract.Linux.gz&rettype=fasta&retmode=text&tool=edirect&edirect=17.8&edirect_os=Linux&email=davo%40DESKTOP-4CTEMG7.localdomain

HTTP/1.1 400 Bad Request
 WARNING:  FAILURE ( Tue Sep 20 14:16:28 PDT 2022 )
nquire -url https://eutils.ncbi.nlm.nih.gov/entrez/eutils/ efetch.fcgi -db nucleotide -id Please,Unable,chmod,dwn,entrez,entrezdirect,executable.,execute,f,following,ftp.ncbi.nlm.nih.gov,gunzip,locate,nquire,the,to,x,xtract,xtract.Linux,xtract.Linux.gz -rettype fasta -retmode text -tool edirect -edirect 17.8 -edirect_os Linux -email davo@DESKTOP-4CTEMG7.localdomain

Please could anyone help out with this?
Thanks.

**Comment by ialbert** [comment]
it seems like one tool did not get installed properly, `xtract`, type that tool name alone, what do you see?

    xtract

can you tell me the version of research?

    esearch -version

**Reply by akgreat22** [reply]
Thanks. 
When I typed xtract, I got this:

Unable to locate xtract executable. Please execute the following:

  nquire -dwn ftp.ncbi.nlm.nih.gov entrez/entrezdirect xtract.Linux.gz
  gunzip -f xtract.Linux.gz
  chmod +x xtract.Linux


The esearch version is 17.8

**Reply by ialbert** [reply]
you must have installed entrez-direct with a different method not conda or my instructions since conda only has version 16.2

use the other instructions that I mention below

**Reply by akgreat22** [reply]
I actually used the source code mentioned in Biostar handbook 2018

**Reply by ialbert** [reply]
well, alas, as it happens in bioinformatics old code does not work; this is why the handbook needs to be maintained year over year

run the following in your terminal:

       # delete old edirect
       rm -rf ~/edirect
       
      # Install new edirect
       yes | sh -c "$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)"
       
 restart the terminal, I think then it will work
 
 but I will say the code in the 2018 version is probably way outdated,

**Comment by ialbert** [comment]
Another possibility would be to install entrez direct from command line as described here:

https://www.ncbi.nlm.nih.gov/books/NBK179288/

basically executing:

    yes | sh -c "$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)"
    
then open a new terminal and the tool should work

**Reply by ialbert** [reply]
This latter one is now the recommended way to install entrez-direct. The main installation will run this same code.

**Reply by akgreat22** [reply]
Thanks for your quick response Dr Albert. However, after running the command line above, I still got the same thing and I checked the xtract and it was giving the same result. Is there anything I'm doing wrong? Really wondering!
Already I am really getting interested and I hope this will be resolved.

**Reply by ialbert** [reply]
you need to remove the old version of efetch, I suspect that your system is set up such that gets pulled in before the updated version

    which esearch

will show you where it is coming from, it should come from the ~/edirect folder and you need to either remove the old location or edit the .bashrc or bash_profile to remove the PATH operations

in general I would recommend starting with a fresh install, the book has a more streamlined install nowadays

**Reply by akgreat22** [reply]
when I ran "which esearch" i got

/home/davo/src/edirect/esearch

**Comment by akgreat22** [comment]
Thanks, this actually worked.
